faunadb-scala/src/main/scala/faunadb/FaunaClient.scala

```diff
@@ -195,18 +213,31 @@ class FaunaClient private (connection: Connection) {
       result.asInstanceOf[ArrayV].elems
     }
 
-  private def performRequest(body: JsonNode, timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[Value] = {
+  private def performRequestCommon[T](body: JsonNode, timeout: Option[FiniteDuration], handler: HttpResponse[String] => Future[T])(implicit ec: ExecutionContext): Future[T] = {
     val javaTimeout = timeout.map(_.toJava).asJava
     val response: Future[HttpResponse[String]] = connection.post("", body, javaTimeout).toScala
 
     response
       .flatMap {
-        case successResponse if successResponse.statusCode() < 300 => handleSuccessResponse(successResponse)
+        case successResponse if successResponse.statusCode() < 300 => handler.apply(successResponse)
         case errorResponse => handleErrorResponse(errorResponse.statusCode(), errorResponse.body())
       }
       .recoverWith(handleNetworkExceptions)
   }
 
+  private def performRequest(body: JsonNode, timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[Value] = {
+    performRequestCommon(body, timeout, handleSuccessResponse)
+  }
+
+  private def performRequestWithMetrics(body: JsonNode, timeout: Option[FiniteDuration])(implicit ec: ExecutionContext): Future[MetricsResponse] = {
+    performRequestCommon(body, timeout, handleSuccessResponseWithMetrics)
+  }
+
+  private def handleSuccessResponseWithMetrics(response: HttpResponse[String])(implicit ec: ExecutionContext): Future[MetricsResponse] = {
+    val metricsMap = (Metrics.All map (item => item -> response.headers().firstValue(item.toString).asScala)).toMap
```

Comment:
We can avoid allocating an intermediary `List` by using an `Iterator`.
```scala
Metrics.All
  .iterator
  .map (item => item -> response.headers().firstValue(item.toString).asScala)
  .toMap
```

Reply:
remade as you suggested, thanks